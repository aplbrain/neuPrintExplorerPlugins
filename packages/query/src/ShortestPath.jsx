/*
 * Query to find shortest path between two neurons.
 */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import randomColor from 'randomcolor';
import { withRouter } from 'react-router';

import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import TextField from '@material-ui/core/TextField';
import { withStyles } from '@material-ui/core/styles';

import { submit } from 'actions/plugins';
import { getQueryString } from 'helpers/queryString';
import { LoadQueryString, SaveQueryString } from 'helpers/qsparser';
import { setUrlQS } from 'actions/app';

const pluginName = 'ShortestPath';

const styles = theme => ({
  textField: {
    margin: 4,
    display: 'block',
    marginLeft: 'auto',
    marginRight: 'auto'
  },
  formControl: {
    margin: theme.spacing.unit,
  },
  button: {
    margin: 4,
    display: 'block'
  },
  clickable: {
    cursor: 'pointer'
  }
});

class ShortestPath extends React.Component {
  constructor(props) {
    super(props);
    const { urlQueryString } = this.props;
    const initqsParams = {
      bodyId1: '',
      bodyId2: '',
      minWeight: 0
    };
    const qsParams = LoadQueryString(
      `Query:${this.constructor.queryName}`,
      initqsParams,
      urlQueryString
    );

    this.state = {
      qsParams
    };
  }

  static get queryName() {
    return 'Shortest path';
  }

  static get queryCategory() {
    return 'recon';
  }

  static get queryDescription() {
    return 'Find all neurons along the shortest path between two neurons.';
  }

  static get isExperimental() {
    return true;
  }

  processResults = (query, apiResponse) => {
    const data = apiResponse.data.map(row => [row[0], row[1], row[2], row[3], row[4]]);

    return {
      columns: ['bodyId', 'name', 'status', '#pre', '#post'],
      data,
      debug: apiResponse.debug
    };
  };

  // creates query object and sends to callback
  processRequest = () => {
    const { dataSet, actions, history } = this.props;
    const { qsParams } = this.state;
    const { bodyId1, bodyId2, minWeight } = qsParams;

    const shortestPathQuery = `MATCH path=shortestPath((a:\`${dataSet}-Neuron\`{bodyId:${bodyId1}})-[r:ConnectsTo*]-(b:\`${dataSet}-Neuron\`{bodyId:${bodyId2}})) WHERE all(rs in r WHERE rs.weight>=${minWeight}) WITH nodes(path) AS ns UNWIND ns AS n RETURN n.bodyId, n.name, n.status, n.pre, n.post`;
    const query = {
      dataSet,
      cypherQuery: shortestPathQuery,
      visType: 'SimpleTable',
      plugin: pluginName,
      parameters: { dataset: dataSet },
      title: `Neurons along path between ${bodyId1} and ${bodyId2}`,
      menuColor: randomColor({ luminosity: 'light', hue: 'random' }),
      processResults: this.processResults
    };
    actions.submit(query);
    history.push({
      pathname: '/results',
      search: getQueryString()
    });
    return query;
  };

  addBodyId1 = event => {
    const { actions } = this.props;
    const { qsParams } = this.state;
    const oldParams = qsParams;
    oldParams.bodyId1 = event.target.value;
    actions.setURLQs(SaveQueryString(`Query:${this.constructor.queryName}`, oldParams));
    this.setState({
      qsParams: oldParams
    });
  };

  addBodyId2 = event => {
    const { actions } = this.props;
    const { qsParams } = this.state;
    const oldParams = qsParams;
    oldParams.bodyId2 = event.target.value;
    actions.setURLQs(SaveQueryString(`Query:${this.constructor.queryName}`, oldParams));
    this.setState({
      qsParams: oldParams
    });
  };

  addMinWeight = event => {
    const { actions } = this.props;
    const { qsParams } = this.state;
    const oldParams = qsParams;
    oldParams.minWeight = event.target.value;
    actions.setURLQs(SaveQueryString(`Query:${this.constructor.queryName}`, oldParams));
    this.setState({
      qsParams: oldParams
    });
  };

  render() {
    const { isQuerying, classes } = this.props;
    const { qsParams } = this.state;

    return (
      <div>
        <FormControl fullWidth className={classes.formControl}>
          <TextField
            label="Body ID A"
            multiline
            fullWidth
            rows={1}
            value={qsParams.bodyId1}
            rowsMax={1}
            className={classes.textField}
            onChange={this.addBodyId1}
          />
          <TextField
            label="Body ID B"
            multiline
            fullWidth
            rows={1}
            value={qsParams.bodyId2}
            rowsMax={1}
            className={classes.textField}
            onChange={this.addBodyId2}
          />
          <TextField
            label="Minimum weight"
            multiline
            fullWidth
            type="number"
            rows={1}
            value={qsParams.minWeight}
            rowsMax={1}
            className={classes.textField}
            onChange={this.addMinWeight}
          />
        </FormControl>
        <Button
          disabled={isQuerying}
          color="primary"
          variant="contained"
          onClick={this.processRequest}
        >
          Submit
        </Button>
      </div>
    );
  }
}

ShortestPath.propTypes = {
  dataSet: PropTypes.string.isRequired,
  isQuerying: PropTypes.bool.isRequired,
  actions: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired,
  urlQueryString: PropTypes.string.isRequired
};

const ShortestPathState = state => ({
  urlQueryString: state.app.get('urlQueryString'),
  isQuerying: state.query.isQuerying
});

// The submit action which will accept your query, execute it and
// store the results for view plugins to display.
const ShortestPathDispatch = dispatch => ({
  actions: {
    submit: query => {
      dispatch(submit(query));
    },
    setURLQs(querystring) {
      dispatch(setUrlQS(querystring));
    }
  }
});

export default withRouter(
  withStyles(styles, { withTheme: true })(
    connect(
      ShortestPathState,
      ShortestPathDispatch
    )(ShortestPath)
  )
);
