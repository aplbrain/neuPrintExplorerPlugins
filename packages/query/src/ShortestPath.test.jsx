import { ShortestPath } from './ShortestPath';

let wrapper;
let button;
let bodyAField;
let bodyBField;
let minWeightField;
const emptyApiResponse = {
  columns: ['length(path)', 'ids', 'weights'],
  data: [],
  debug: 'test'
};
const apiResponse = {
  columns: ['length(path)', 'ids', 'weights'],
  data: [
    [3, [[758065781, 'a'], [5813020577, 'b'], [481112669, 'c'], [8000, 'd']], [3, 5, 9]],
    [3, [[758065781, 'a'], [5813020577, 'b'], [9000, 'e'], [8000, 'd']], [3, 4, 6]]
  ],
  debug: 'test'
};

const styles = { select: {}, clickable: {} };
const { actions, React, enzyme, renderer } = global;

const neoServerSettings = {
  get: () => 'http://example.com'
};

const component = (
  <ShortestPath
    availableROIs={['roiA', 'roiB', 'roiC']}
    dataSet="test"
    datasetstr="test"
    actions={actions}
    classes={styles}
    history={{ push: jest.fn() }}
    isQuerying={false}
    neoServerSettings={neoServerSettings}
    neoServer="testServer"
  />
);

describe('shortest path Plugin', () => {
  beforeAll(() => {
    wrapper = enzyme.mount(component);
    button = wrapper.find('Button');
    bodyAField = wrapper.find('TextField').at(0);
    bodyBField = wrapper.find('TextField').at(1);
    minWeightField = wrapper.find('TextField').at(2);
  });
  beforeEach(() => {
    actions.submit.mockClear();
  });
  it('has name and description', () => {
    expect(ShortestPath.queryName).toBeTruthy();
    expect(ShortestPath.queryDescription).toBeTruthy();
  });
  it('renders correctly', () => {
    const pluginView = renderer.create(component).toJSON();
    expect(pluginView).toMatchSnapshot();
  });
  describe('when user clicks submit', () => {
    it('should return a query object with input fields contained in cypher string and should submit', () => {
      bodyAField.props().onChange({ target: { value: '123456' } });
      bodyBField.props().onChange({ target: { value: '645321' } });
      minWeightField.props().onChange({ target: { value: '6' } });

      const shortestPathQuery =
        "MATCH (a:`test-Neuron`{bodyId:123456}), (b:`test-Neuron`{bodyId:645321}) CALL analysis.getShortestPathWithMinWeight(a, b, 'ConnectsTo>', 'prop', 'weight', 1, 6) YIELD path,weight WITH extract(n IN nodes(path) | [n.bodyId,n.name]) AS ids,extract(rst IN rels(path) | rst.weight) AS weights, path RETURN length(path), ids, weights";

      expect(button.props().onClick()).toEqual(
        expect.objectContaining({
          dataSet: 'test',
          cypherQuery: shortestPathQuery,
          visType: 'Graph',
          plugin: 'ShortestPath',
          parameters: { dataset: 'test' },
          title: expect.any(String),
          menuColor: expect.any(String),
          processResults: expect.any(Function)
        })
      );
      expect(actions.submit).toHaveBeenCalledTimes(1);
    });
  });

  describe('processes returned results', () => {
    it('should produce error, empty data object if results are empty', () => {
      const processedEmptyResults = wrapper.instance().processResults({}, emptyApiResponse);
      expect(actions.pluginResponseError).toHaveBeenCalledTimes(1);
      expect(processedEmptyResults).toEqual({
        columns: emptyApiResponse.columns,
        data: emptyApiResponse.data,
        debug: emptyApiResponse.debug
      });
    });
    it('should produce object with raw neo4j results as "data" (for csv) and graph nodes and edges as "graph" (for cytoscape)', () => {
      const processedResults = wrapper.instance().processResults({}, apiResponse);
      const { columns, data, graph, debug } = processedResults;
      expect(columns).toBe(apiResponse.columns);
      expect(data).toBe(apiResponse.data);
      expect(debug).toBe(apiResponse.debug);
      const { elements, minWeight, maxWeight } = graph;
      expect(minWeight).toBe(3);
      expect(maxWeight).toBe(9);
      const { nodes, edges } = elements;
      expect(nodes.length).toBe(5);
      expect(edges.length).toBe(5);
    });
  });
});
