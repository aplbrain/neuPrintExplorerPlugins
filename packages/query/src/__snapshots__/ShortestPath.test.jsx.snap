// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shortest path Plugin renders correctly 1`] = `
<div>
  <div
    className="MuiFormControl-root-1 MuiFormControl-fullWidth-4"
  >
    <div
      className="MuiFormControl-root-1 MuiFormControl-fullWidth-4"
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
        data-shrink={false}
      >
        Body ID A
      </label>
      <div
        className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-fullWidth-45 MuiInput-fullWidth-30 MuiInputBase-formControl-37 MuiInput-formControl-24 MuiInputBase-multiline-44 MuiInput-multiline-29"
        onClick={[Function]}
      >
        <div
          className="MuiPrivateTextarea-root-53"
        >
          <textarea
            aria-hidden="true"
            className="MuiPrivateTextarea-textarea-54 MuiPrivateTextarea-shadow-55"
            readOnly={true}
            rows="1"
            tabIndex={-1}
            value=""
          />
          <textarea
            aria-hidden="true"
            className="MuiPrivateTextarea-textarea-54 MuiPrivateTextarea-shadow-55"
            readOnly={true}
            rows={1}
            tabIndex={-1}
            value=""
          />
          <textarea
            aria-invalid={false}
            className="MuiPrivateTextarea-textarea-54 MuiInputBase-input-46 MuiInput-input-31 MuiInputBase-inputMultiline-48 MuiInput-inputMultiline-33"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            rows={1}
            style={
              Object {
                "height": 19,
              }
            }
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="MuiFormControl-root-1 MuiFormControl-fullWidth-4"
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
        data-shrink={false}
      >
        Body ID B
      </label>
      <div
        className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-fullWidth-45 MuiInput-fullWidth-30 MuiInputBase-formControl-37 MuiInput-formControl-24 MuiInputBase-multiline-44 MuiInput-multiline-29"
        onClick={[Function]}
      >
        <div
          className="MuiPrivateTextarea-root-53"
        >
          <textarea
            aria-hidden="true"
            className="MuiPrivateTextarea-textarea-54 MuiPrivateTextarea-shadow-55"
            readOnly={true}
            rows="1"
            tabIndex={-1}
            value=""
          />
          <textarea
            aria-hidden="true"
            className="MuiPrivateTextarea-textarea-54 MuiPrivateTextarea-shadow-55"
            readOnly={true}
            rows={1}
            tabIndex={-1}
            value=""
          />
          <textarea
            aria-invalid={false}
            className="MuiPrivateTextarea-textarea-54 MuiInputBase-input-46 MuiInput-input-31 MuiInputBase-inputMultiline-48 MuiInput-inputMultiline-33"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            rows={1}
            style={
              Object {
                "height": 19,
              }
            }
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="MuiFormControl-root-1 MuiFormControl-fullWidth-4"
    >
      <label
        className="MuiFormLabel-root-16 MuiFormLabel-filled-20 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
        data-shrink={true}
      >
        Minimum weight
      </label>
      <div
        className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-fullWidth-45 MuiInput-fullWidth-30 MuiInputBase-formControl-37 MuiInput-formControl-24 MuiInputBase-multiline-44 MuiInput-multiline-29"
        onClick={[Function]}
      >
        <div
          className="MuiPrivateTextarea-root-53"
        >
          <textarea
            aria-hidden="true"
            className="MuiPrivateTextarea-textarea-54 MuiPrivateTextarea-shadow-55"
            readOnly={true}
            rows="1"
            tabIndex={-1}
            value=""
          />
          <textarea
            aria-hidden="true"
            className="MuiPrivateTextarea-textarea-54 MuiPrivateTextarea-shadow-55"
            readOnly={true}
            rows={1}
            tabIndex={-1}
            value={0}
          />
          <textarea
            aria-invalid={false}
            className="MuiPrivateTextarea-textarea-54 MuiInputBase-input-46 MuiInput-input-31 MuiInputBase-inputType-49 MuiInput-inputType-34 MuiInputBase-inputMultiline-48 MuiInput-inputMultiline-33"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            rows={1}
            style={
              Object {
                "height": 19,
              }
            }
            value={0}
          />
        </div>
      </div>
    </div>
  </div>
  <button
    className="MuiButtonBase-root-82 MuiButton-root-56 MuiButton-contained-67 MuiButton-containedPrimary-68 MuiButton-raised-70 MuiButton-raisedPrimary-71"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-57"
    >
      Submit
    </span>
    <span
      className="MuiTouchRipple-root-85"
    />
  </button>
</div>
`;
